File: samples/tutorials-native/tutorials-2d/tutorial-4/javaBuilder/src/main/java/com/scichart/tutorial/MainActivity.java
package com.scichart.tutorial;

import android.os.Bundle;
import android.util.Log;
import android.view.Gravity;
import android.widget.LinearLayout;

import androidx.appcompat.app.AppCompatActivity;

import com.scichart.charting.model.dataSeries.XyDataSeries;
import com.scichart.charting.modifiers.ModifierGroup;
import com.scichart.charting.modifiers.PinchZoomModifier;
import com.scichart.charting.modifiers.RolloverModifier;
import com.scichart.charting.modifiers.ZoomExtentsModifier;
import com.scichart.charting.modifiers.ZoomPanModifier;
import com.scichart.charting.visuals.SciChartSurface;
import com.scichart.charting.visuals.axes.IAxis;
import com.scichart.charting.visuals.pointmarkers.EllipsePointMarker;
import com.scichart.charting.visuals.renderableSeries.IRenderableSeries;
import com.scichart.core.annotations.Orientation;
import com.scichart.core.framework.UpdateSuspender;
import com.scichart.core.model.DoubleValues;
import com.scichart.core.model.IntegerValues;
import com.scichart.extensions.builders.SciChartBuilder;

import java.util.Collections;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

public class MainActivity extends AppCompatActivity {
    // <DataSeriesSetup>
    private int pointsCount = 200;

    private ScheduledExecutorService scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
    private ScheduledFuture schedule;

    private SciChartSurface surface;

    private final DoubleValues lineData = new DoubleValues();
    private XyDataSeries lineDataSeries;

    private DoubleValues scatterData = new DoubleValues();
    private XyDataSeries scatterDataSeries;
    // </DataSeriesSetup>

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        try {
            SciChartSurface.setRuntimeLicenseKey("");
        } catch (Exception e) {
            Log.e("SciChart", "Error during setting license key", e);
        }

        surface = new SciChartSurface(this);
        LinearLayout chartLayout = findViewById(R.id.chart_layout);
        chartLayout.addView(surface);

        SciChartBuilder.init(this);
        final SciChartBuilder sciChartBuilder = SciChartBuilder.instance();

        final IAxis xAxis = sciChartBuilder.newNumericAxis()
                .withAxisTitle("X Axis Title")
                .build();

        final IAxis yAxis = sciChartBuilder.newNumericAxis()
                .withAxisTitle("Y Axis Title")
                .build();

        UpdateSuspender.using(surface, () -> {
            Collections.addAll(surface.getXAxes(), xAxis);
            Collections.addAll(surface.getYAxes(), yAxis);
        });

        // <SetFifoCapacity>
        lineDataSeries = sciChartBuilder.newXyDataSeries(Integer.class, Double.class)
                .withSeriesName("Line Series")
                .withFifoCapacity(300)
                .build();
        scatterDataSeries = sciChartBuilder.newXyDataSeries(Integer.class, Double.class)
                .withSeriesName("Scatter Series")
                .withFifoCapacity(300)
                .build();
        // </SetFifoCapacity>

        // <InitialDataSetup>
        final IntegerValues xValues = new IntegerValues();
        for (int i = 0; i < pointsCount; i++) {
            xValues.add(i);
            lineData.add(Math.sin(i * 0.1));
            scatterData.add(Math.cos(i * 0.1));
            count += 1;
        }
        lineDataSeries.append(xValues, lineData);
        scatterDataSeries.append(xValues, scatterData);
        // </InitialDataSetup>

        final IRenderableSeries lineSeries = sciChartBuilder.newLineSeries()
                .withDataSeries(lineDataSeries)
                .build();

        final EllipsePointMarker pointMarker = sciChartBuilder
                .newPointMarker(new EllipsePointMarker())
                .withFill(0xFF32CD32)
                .withSize(10)
                .build();

        final IRenderableSeries scatterSeries = sciChartBuilder.newScatterSeries()
                .withDataSeries(scatterDataSeries)
                .withPointMarker(pointMarker)
                .build();

        ModifierGroup modifierGroup = sciChartBuilder.newModifierGroup()
                .withLegendModifier()
                .withPosition(Gravity.BOTTOM | Gravity.CENTER_HORIZONTAL, 10)
                .withOrientation(Orientation.HORIZONTAL)
                .build()
                .build();

        UpdateSuspender.using(surface, () -> {
            Collections.addAll(surface.getRenderableSeries(), lineSeries, scatterSeries);
            Collections.addAll(surface.getChartModifiers(), new PinchZoomModifier(), new ZoomPanModifier(), new ZoomExtentsModifier());
            Collections.addAll(surface.getChartModifiers(), modifierGroup);
            Collections.addAll(surface.getChartModifiers(), new RolloverModifier());
        });

        // <CreateTimer>
        schedule = scheduledExecutorService.scheduleWithFixedDelay(updateData, 0, 10, TimeUnit.MILLISECONDS);
        // </CreateTimer>
    }
/*
    // <UpdateData>
    private Double phase = 0.0;

    private final Runnable updateData = () -> {
        for (int i = 0; i < pointsCount; i++) {
            lineData.set(i, Math.sin(i* 0.1 + phase));
            scatterData.set(i, Math.cos(i* 0.1 + phase));
        }

        UpdateSuspender.using(surface, () -> {
            lineDataSeries.updateRangeYAt(0, lineData);
            scatterDataSeries.updateRangeYAt(0, scatterData);

            // zoom series to fit viewport size into X-Axis direction
            surface.zoomExtentsX();
        });

        phase += 0.01;
    };
    // </UpdateData>
 */

    // <AppendData>
    private Integer count = 0;

    private final Runnable updateData = () -> {
        Integer x = count;
        UpdateSuspender.using(surface, () -> {
            lineDataSeries.append(x, Math.sin(x* 0.1));
            scatterDataSeries.append(x, Math.cos(x * 0.1));

            // zoom series to fit viewport size into X-Axis direction
            surface.zoomExtentsX();
            count += 1;
        });
    };
    // </AppendData>
}

File: samples/tutorials-xamarin/tutorials-2d/tutorial-04/MainActivity.cs
using System;
using Android.App;
using Android.OS;
using Android.Views;
using AndroidX.AppCompat.App;
using SciChart.Core.Framework;
using SciChart.Charting.Visuals;
using SciChart.Charting.Visuals.Axes;
using SciChart.Charting.Model.DataSeries;
using SciChart.Charting.Visuals.RenderableSeries;
using SciChart.Charting.Visuals.PointMarkers;
using SciChart.Drawing.Common;
using SciChart.Charting.Modifiers;
using System.Timers;
using SciChart.Core.Model;

namespace Tutorial_2d_04
{
    [Activity(Label = "@string/app_name", Theme = "@style/AppTheme.NoActionBar", MainLauncher = true)]
    public class MainActivity : AppCompatActivity
    {
        private readonly object _syncRoot = new object();

        // <DataSeriesSetup>
        private const int PointsCount = 200;
        private const long TimerInterval = 10;

        private Timer _timer;

        private SciChartSurface surface;

        private readonly DoubleValues lineData = new DoubleValues();
        private readonly XyDataSeries<int, double> lineDataSeries = new XyDataSeries<int, double>() { SeriesName = "Line Series" };

        private readonly DoubleValues scatterData = new DoubleValues();
        private readonly XyDataSeries<int, double> scatterDataSeries = new XyDataSeries<int, double>() { SeriesName = "Scatter Series" };
        // </DataSeriesSetup>

        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            Xamarin.Essentials.Platform.Init(this, savedInstanceState);

            SciChartSurface.SetRuntimeLicenseKey("");
            SetContentView(Resource.Layout.activity_main);

            surface = new SciChartSurface(this);
            var chartLayout = FindViewById<ViewGroup>(Resource.Id.chart_layout);
            chartLayout.AddView(surface);

            // <SetFifoCapacity>
            lineDataSeries.FifoCapacityValue = 300;
            scatterDataSeries.FifoCapacityValue = 300;
            // </SetFifoCapacity>

            // <InitialDataSetup>
            var xValues = new IntegerValues();

            for (int i = 0; i < PointsCount; i++) 
            {
                xValues.Add(i);
                lineData.Add(Math.Sin(i * 0.1));
                scatterData.Add(Math.Cos(i * 0.1));
                count += 1;
            }  
            lineDataSeries.Append(xValues, lineData);
            scatterDataSeries.Append(xValues, scatterData);
            // </InitialDataSetup>

            var lineSeries = new FastLineRenderableSeries() { DataSeries = lineDataSeries };
            var scatterSeries = new XyScatterRenderableSeries()
            {
                DataSeries = scatterDataSeries,
                PointMarker = new EllipsePointMarker()
                {
                    Width = 10,
                    Height = 10,
                    FillStyle = new SolidBrushStyle(0xFF32CD32)
                }
            };

            // <CreateLegend>
            var legendModifier = new LegendModifier(this);
            legendModifier.SetOrientation(Orientation.Horizontal);
            legendModifier.SetLegendPosition(GravityFlags.Bottom | GravityFlags.CenterHorizontal, 0, 0, 0, 10);
            // </CreateLegend>

            using (surface.SuspendUpdates())
            {
                surface.XAxes.Add(new NumericAxis(this) { AxisTitle = "X Axis Title" } );
                surface.YAxes.Add(new NumericAxis(this) { AxisTitle = "Y Axis Title" } );

                surface.RenderableSeries.Add(lineSeries);
                surface.RenderableSeries.Add(scatterSeries);

                surface.ChartModifiers.Add(new PinchZoomModifier());
                surface.ChartModifiers.Add(new ZoomPanModifier());
                surface.ChartModifiers.Add(new ZoomExtentsModifier());
                surface.ChartModifiers.Add(legendModifier);
                surface.ChartModifiers.Add(new RolloverModifier());
            }

            // <CreateTimer>
            _timer = new Timer(TimerInterval) { AutoReset = true };
            _timer.Elapsed += OnTick;
            _timer.Start();
            // </CreateTimer>
        }

        // <AppendData>
        private int count = 0;

        private void OnTick(object sender, ElapsedEventArgs e)
        {
            lock (_syncRoot)
            {
                var x = count;
                using (surface.SuspendUpdates())
                {
                    lineDataSeries.Append(x, Math.Sin(x * 0.1));
                    scatterDataSeries.Append(x, Math.Cos(x * 0.1));

                    count += 1;
                    surface.ZoomExtentsX();
                }
            }
        }
        // </AppendData>

        /*
        // <UpdateData>
        private Double phase = 0.0;

        private void OnTick(object sender, ElapsedEventArgs e)
        {
            lock (_syncRoot)
            {
                for (int i = 0; i < PointsCount; i++)
                {
                    lineData.Set(i, Math.Sin(i * 0.1 + phase));
                    scatterData.Set(i, Math.Cos(i * 0.1 + phase));
                }
                
                using (surface.SuspendUpdates())
                {
                    lineDataSeries.UpdateRangeYAt(0, lineData);
                    scatterDataSeries.UpdateRangeYAt(0, scatterData);

                    // zoom series to fit viewport size into X-Axis direction
                    surface.ZoomExtentsX();
                }

                phase += 0.01;
            }
        }
        // </UpdateData>
        */
	}
}